library/src/main/java/org/apache/apex/malhar/lib/dedup/AbstractDeduper.java

```diff
@@ -0,0 +1,462 @@
+      recordDecision(tuple, Decision.UNKNOWN);
+    }
+  }
+
+  /**
+   * Processes a valid (non-expired) tuple. This tuple may be a unique or a duplicate.
+   *
+   * @param tuple
+   *          The tuple to be processed
+   * @param value
+   *          Looked up key of the tuple
+   */
+  protected void processValid(T tuple, Slice value)
+  {
+    if (!orderedOutput || waitingEvents.isEmpty()) {
+      if (value == null) {
+        managedState.put(getTime(tuple), getKey(tuple), getKey(tuple));
+        processUnique(tuple);
+      } else {
+        processDuplicate(tuple);
+      }
+    } else {
+      processWaitingEvent(tuple, Futures.immediateFuture(value));
+    }
+  }
+
+  /**
+   * Processes invalid tuples.
+   *
+   * @param tuple
+   */
+  protected void processInvalid(T tuple)
+  {
+    if (orderedOutput && !decisions.isEmpty()) {
+      recordDecision(tuple, Decision.EXPIRED);
+    } else {
+      processExpired(tuple);
+    }
+  }
+
+  /**
+   * Processes an expired tuple
+   *
+   * @param tuple
+   */
+  protected void processExpired(T tuple)
+  {
+    expiredEvents++;
+    emitExpired(tuple);
+  }
+
+  /**
+   * Processes the duplicate tuple.
+   *
+   * @param tuple
+   *          The tuple which is a duplicate
+   */
+  protected void processDuplicate(T tuple)
+  {
+    if (orderedOutput && !decisions.isEmpty()) {
+      recordDecision(tuple, Decision.DUPLICATE);
+    } else {
+      duplicateEvents++;
+      emitDuplicate(tuple);
+    }
+  }
+
+  /**
+   * Processes the unique tuple.
+   *
+   * @param tuple
+   *          The tuple which is a unique
+   */
+  protected void processUnique(T tuple)
+  {
+    if (orderedOutput && !decisions.isEmpty()) {
+      recordDecision(tuple, Decision.UNIQUE);
+    } else {
+      uniqueEvents++;
+      emitOutput(tuple);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void handleIdleTime()
+  {
+    if (orderedOutput) {
+      emitProcessedTuples();
+    }
+    processAuxiliary(false);
+  }
+
+  /**
+   * Does any auxiliary processing in the idle time of the operator.
+   * Processes any tuples which are waiting for the lookup to return.
+   *
+   * @param finalize Whether or not to wait for future to return
+   */
+  protected void processAuxiliary(boolean finalize)
+  {
+    if (waitingEvents.size() > 0) {
+      Iterator<Map.Entry<T, Future<Slice>>> waitIterator = waitingEvents.entrySet().iterator();
+      while (waitIterator.hasNext()) {
+        Map.Entry<T, Future<Slice>> waitingEvent = waitIterator.next();
+        T tuple = waitingEvent.getKey();
+        Slice tupleKey = getKey(tuple);
+        long tupleTime = getTime(tuple);
+        Future<Slice> future = waitingEvent.getValue();
+        if (future.isDone() || finalize ) {
+          try {
+            if (future.get() == null && asyncEvents.get(tupleKey) == null) {
+              managedState.put(tupleTime, tupleKey, tupleKey);
+              asyncEvents.put(tupleKey, tupleTime);
```

Comment:
I don't see value present in asyncEvents being used anywhere.. Can you put NULL Object instead?